src/main/scala/org/ergoplatform/utils/metrics.scala

```diff
@@ -0,0 +1,296 @@
+package org.ergoplatform.utils
+
+import java.io.{BufferedWriter, FileWriter, File, Writer}
+
+import org.ergoplatform.settings.ErgoSettings
+import scalan.util.FileUtil
+
+import scala.util.{DynamicVariable, Try}
+import scorex.core.validation.ValidationResult
+import scorex.util.ModifierId
+
+import scala.collection.mutable
+
+/** Implementation of the metrics sub-system. The general algorithm of metrics collection is as the
+  * following:
+  * 1. The `block` of code is wrapped into invocation of [[metrics.measureOp]] or
+  * [[metrics.measureCostedOp]] method.
+  * 2. The method executes the `block` and if `ergo.node.collectMetrics` config parameter is `true`
+  * then execution time is measured.
+  * 3. If metric collection is enabled then [[metrics.MeasuredData]] instance is created and passed
+  * to the current collector on the thread [[metrics.MetricsCollector.current]] by calling
+  * [[metrics.MetricsCollector.collect]] method.
+  * 4. The [[metrics.CsvCollector]] assembles a new csv line, looks up the
+  * [[metrics.CsvCollector.OutputManager]], and writes the line into the output Writer.
+  *
+  * The current collector on the current thread can be changed by using the following pattern
+  * <pre class="stHighlight">
+  * metrics.executeWithCollector(MyCollector) {
+  *   // execute code with the given collector set as current
+  * }
+  * </pre>
+  */
+object metrics {
+
+  /** Type-class which implements capability to output data as CSV-like row. */
+  trait MetricRow[D] {
+    /** Returns an array of columns in a CSV-like table */
+    def fieldNames: Seq[String]
+    /** Returns a CSV-like row for the given data instance.
+      * @param d an instance to transform into a row of strings
+      * @return an array of cells in a CSV-like row
+      */
+    def fieldValues(d: D): Seq[String]
+  }
+
+  /** Represents a descriptor for the given metric name and data type.
+    * Conceptually a descriptor specifies is a table of rows with the set of columns given by `row`
+    * instance.
+    * @tparam D data type which represents information recorded for each measurement of this metric
+    * @param metricName a name of a metric
+    * @param row an instance of CSV-like row handler
+    */
+  case class MetricDesc[D](metricName: String)(implicit row: MetricRow[D]) {
+    /** If metric is a table, then this method returns a the names of the columns. */
+    def fieldNames: Seq[String] = row.fieldNames
+    /** Extract the values of the row from the given data instance. */
+    def fieldValues(d: D): Seq[String] = row.fieldValues(d)
+  }
+
+  val emptyModifierId: ModifierId = ModifierId @@ ""
```

Comment:
ModifierId is supposed to be Base16 encoded 256-bit string, it is not compile-checked as compile-time checks are hard (especially in Scala), but do not break semantics in the runtime because of that! Use some special value, or better, optional type when needed.

Reply:
removed and used Optional type.